### ISSUE TITLE
DDLDatabase loads too many files when using relative path

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#8335]
When specifying a relative path like this:

```xml
<database>
	<name>org.jooq.meta.extensions.ddl.DDLDatabase</name>
	<properties>
		<property>
			<key>scripts</key>
			<value>src/main/resourc*/**/*.sql</value>
		</property>
	</properties>
</database>
```

The `DDLDatabase` is loading too many files depending on how it is run.

This is because of #7767. When running the DDLDatabase from within a module, the current directory is the same as the Maven `${basedir}`. But when running it from a global build (the parent module), then the current directory is different, and more of the path will be scanned.

Related: #2887

**Comment by lukaseder** [comment]
The name resolution is not directly related to Maven. What we need here is a way for the code generator to resolve names generically, regardless who is running it, i.e. a `basedir` variable.

The Maven plugin would then specify that variable to be the same as the `${basedir}` of the module.

**Comment by knutwannheden** [comment]
Another problem (also solved by having an explicit `basedir`) is when using `<configurationFile>`: The external configuration file cannot use `${basedir}` for the `scripts` value to make paths absolute, since this isn't a Maven `pom.xml` file. Thus relative paths will here always be resolved to the current working directory.

**Comment by knutwannheden** [comment]
We should also consider backporting the fix to 3.12.

**Comment by lukaseder** [comment]
I think it might be worth solving this more generically in the core library as we might be able to re-use it across various locations that work with the new `org.jooq.Source` type, which will be introduced in 3.13 (#9043).

I'm thinking of moving the `org.jooq.meta.tools.FilePattern` API to the core, refactoring it to make it more generally useful. I'll document the functionality in a new issue that I'll create right away.

> We should also consider backporting the fix to 3.12.

I don't think this can be solved thoroughly with a "fix". It is a complex problem with many aspects. I'm a bit uneasy with a quick fix that works in Maven, but not with `<configurationFile>`, nor with standalone code generation using `GenerationTool`.

We'll be releasing 3.13 in Q1 2020, so I think that prioritising this for 3.13 should be sufficient.

**Comment by lukaseder** [comment]
The new issue is: #9725

**Comment by lukaseder** [comment]
The prerequisites for `${basedir}` support are now implemented in #9725. There are a few issues that are due to bad `FilePattern` implementation caused by a previous fix: https://github.com/jOOQ/jOOQ/issues/8336. These will be addressed in another, new issue first: https://github.com/jOOQ/jOOQ/issues/9726

**Comment by lukaseder** [comment]
> I'm a bit uneasy with a quick fix that works in Maven, but not with `<configurationFile>`, nor with standalone code generation using `GenerationTool`.

In fact, my fear here was unjustified. `<configurationFile>` is a Maven plugin only feature. It does not exist as a standalone feature.

Here's what could be done. There could be a `configuration/basedir` element in the XML configuration, which also works with Maven configurations. That `basedir` would, by default, correspond to the Maven basedir when using the Maven plugin. I'll experiment with that and see how this fixes the regression introduced in https://github.com/jOOQ/jOOQ/issues/9726#issuecomment-574188556

**Comment by lukaseder** [comment]
The new `basedir` configuration will be implemented through #9727
